lighthouse-core/gather/gatherers/image-usage.js

```diff
@@ -0,0 +1,150 @@
+    if (!srcset) {
+      return [];
+    }
+
+    const entries = srcset.split(',');
+    const relativeUrls = entries.map(entry => entry.trim().split(' ')[0]);
+    return relativeUrls.map(url => {
+      const a = document.createElement('a');
+      a.href = url;
+      return a.href;
+    });
+  }
+
+  function toObject(tag) {
+    return {
+      tagName: tag.tagName,
+      src: tag.currentSrc,
+      srcset: tag.srcset,
+      srcsetUrls: parseSrcsetUrls(tag.srcset),
+      sizes: tag.sizes,
+      media: tag.media,
+      clientWidth: tag.clientWidth,
+      clientHeight: tag.clientHeight,
+      naturalWidth: tag.naturalWidth,
+      naturalHeight: tag.naturalHeight,
+    };
+  }
+
+  return [...document.querySelectorAll('img')].map(tag => {
+    if (tag.parentElement.tagName !== 'PICTURE') {
+      return Object.assign(toObject(tag), {isPicture: false});
+    }
+
+    // `img` tags within `picture` behave strangely
+    // The picture's width and height are reflected by the img tag's width
+    // and height but the natural size is not necessarily accurate.
+    const imgTagInfo = toObject(tag);
+    const sources = [...tag.parentElement.children]
+        .filter(tag => tag.tagName === 'SOURCE')
+        .filter(tag => !tag.media || window.matchMedia(tag.media).matches)
+        .map(toObject)
+        .concat(imgTagInfo);
+    return Object.assign(imgTagInfo, {
+      isPicture: true,
+      // nested chain is too deep for DevTools to handle so stringify
+      sources: JSON.stringify(sources),
+    });
+  });
+}
+
+/* istanbul ignore next */
+function determineNaturalSize(url) {
+  return new Promise((resolve, reject) => {
+    const img = new Image();
+    img.addEventListener('error', reject);
+    img.addEventListener('load', () => {
+      resolve({
+        naturalWidth: img.naturalWidth,
+        naturalHeight: img.naturalHeight
+      });
+    });
+
+    img.src = url;
+  });
+}
+
+function pick(obj, keys) {
+  const out = {};
+  for (const key of keys) {
+    out[key] = obj[key];
+  }
+  return out;
+}
+
+class ImageUsage extends Gatherer {
+
+  /**
+   * @param {!{src: string}} tag
+   * @return {Promise<!Object>}
+   */
+  fetchTagWithSizeInformation(tag) {
+    const url = JSON.stringify(tag.src);
+    return this.driver.evaluateAsync(`(${determineNaturalSize.toString()})(${url})`)
+      .then(size => {
+        return Object.assign(tag, size);
+      });
+  }
+
+  afterPass(options, traceData) {
+    const driver = this.driver = options.driver;
+    const indexedNetworkRecords = traceData.networkRecords.reduce((map, record) => {
+      if (/^image/.test(record._mimeType)) {
+        map[record._url] = pick(record, [
+          'url', 'requestHeaders', 'resourceSize',
+          'startTime', 'endTime',
+        ]);
+      }
+
+      return map;
+    }, {});
+
+    return driver.evaluateAsync(`(${collectImageTagInfo.toString()})()`)
+      .then(tags => {
+        return Promise.all(tags.map(tag => {
```

Comment:
> there shouldn't be many network requests happening here and if there end up being many we should revisit

yeah, I'm mostly worried about the cache behavior on devices with less storage. It's possible that for an image on a page the decoded pixel data is still present but the image file has been evicted from the network cache, so `img.src = url` may force another download. We'll have to investigate to find out if that is possible.